x-pack/plugins/apm/server/routes/agent_explorer/get_agents_items.ts

```diff
@@ -0,0 +1,147 @@
+    apm: {
+      events: [ProcessorEvent.metric],
+    },
+    body: {
+      track_total_hits: false,
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            {
+              exists: {
+                field: AGENT_NAME,
+              },
+            },
+            {
+              exists: {
+                field: AGENT_VERSION,
+              },
+            },
+            ...rangeQuery(start, end),
+            ...environmentQuery(environment),
+            ...kqlQuery(kuery),
+            ...(serviceName ? termQuery(SERVICE_NAME, serviceName) : []),
+            ...(agentLanguage
+              ? termQuery(SERVICE_LANGUAGE_NAME, agentLanguage)
+              : []),
+          ],
+        },
+      },
+      aggs: {
+        sample: {
+          random_sampler: randomSampler,
+          aggs: {
+            services: {
+              terms: {
+                field: SERVICE_NAME,
+                size: MAX_NUMBER_OF_SERVICES,
+              },
+              aggs: {
+                instances: {
+                  cardinality: {
+                    field: SERVICE_NODE_NAME,
+                  },
+                },
+                serviceNodes: {
+                  terms: {
+                    field: SERVICE_NODE_NAME,
+                    size: MAX_NUMBER_OF_SERVICES,
+                  },
+                  aggs: {
+                    sample: {
+                      top_metrics: {
+                        metrics: [
+                          { field: AGENT_NAME } as const,
+                          { field: AGENT_VERSION } as const,
+                          { field: SERVICE_LANGUAGE_NAME } as const,
+                        ],
+                        sort: {
+                          '@timestamp': 'desc' as const,
+                        },
+                      },
+                    },
+                  },
+                },
+                environments: {
+                  terms: {
+                    field: SERVICE_ENVIRONMENT,
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  return (
+    response.aggregations?.sample.services.buckets.map((bucket) => {
+      const agents = bucket.serviceNodes.buckets.map((serviceNode) => ({
+        name: serviceNode.sample.top[0].metrics[AGENT_NAME] as AgentName,
+        version: serviceNode.sample.top[0].metrics[AGENT_VERSION] as string,
+      }));
+
+      return {
+        serviceName: bucket.key as string,
+        environments: bucket.environments.buckets.map(
+          (env) => env.key as string
+        ),
+        agentName: map(agents, 'name')?.[0],
+        agentVersion: uniq(map(agents, 'version')),
+        // service.node.name is set by the server only if a container.id or host.name are set. Otherwise should be explicitly set by agents.
+        instances: (bucket.instances.value as number) || 0,
```

Comment:
I think this should always be 1 no?

Reply:
If we default this to 1, what will happen when we query for `service.node.name` and we get no results there?
wouldn't that be inconsistent?

Reply:
https://user-images.githubusercontent.com/1313018/200026697-1a8f6487-b4d0-450a-b882-7779f20db86e.mov